src/System.IO.Compression.Brotli/System/IO/Compression/Brotli.cs

```diff
@@ -4,15 +4,116 @@
-        const int DefaultWindowSize = 24;
+        private const int MinWindowBits = 10;
+        private const int MaxWindowBits = 24;
+        private const int MinQuality = 0;
+        private const int MaxQuality = 11;
+
+        public struct State : IDisposable
+        {
+            internal IntPtr BrotliNativeState { get; private set; }
+            internal BrotliDecoderResult LastDecoderResult;
+            public bool CompressMode { get; private set; }
+
+            public void Dispose()
+            {
+                if (CompressMode)
+                {
+                    BrotliNative.BrotliEncoderDestroyInstance(BrotliNativeState);
+                }
+                else
+                {
+                    BrotliNative.BrotliDecoderDestroyInstance(BrotliNativeState);
+                }
+            }
+
+            internal void InitializeDecoder()
+            {
+                BrotliNativeState = BrotliNative.BrotliDecoderCreateInstance();
+                LastDecoderResult = BrotliDecoderResult.NeedsMoreInput;
+                if (BrotliNativeState == IntPtr.Zero)
+                {
+                    throw new System.Exception(BrotliEx.DecoderInstanceCreate);
+                }
+                CompressMode = false;
+            }
+
+            internal void InitializeEncoder()
+            {
+                
+                BrotliNativeState = BrotliNative.BrotliEncoderCreateInstance();
+                if (BrotliNativeState == IntPtr.Zero)
+                {
+                    throw new System.Exception(BrotliEx.EncoderInstanceCreate);
+                }
+                CompressMode = true;
+            }
+
+            public void SetQuality(uint quality)
+            {
+                if (BrotliNativeState == IntPtr.Zero)
+                {
+                    InitializeEncoder();
+                }
+                if (quality > MaxQuality)
+                {
+                    throw new ArgumentException(BrotliEx.WrongQuality);
+                }
+                BrotliNative.BrotliEncoderSetParameter(BrotliNativeState, BrotliEncoderParameter.Quality, quality);
+            }
+
+            public void SetQuality()
+            {
+                SetQuality(MaxQuality);
+            }
+
+            public void SetWindow(uint window)
+            {
+                if (BrotliNativeState == IntPtr.Zero)
+                {
+                    InitializeEncoder();
+                }
+                if (window - MinWindowBits > MaxWindowBits - MinWindowBits)
+                {
+                    throw new ArgumentException(BrotliEx.WrongWindowSize);
+                }
+                BrotliNative.BrotliEncoderSetParameter(BrotliNativeState, BrotliEncoderParameter.LGWin, window);
+            }
+
+            public void SetWindow()
+            {
+                SetWindow(MaxWindowBits);
+            }
+        }
+
+        public static void EnsureInitialized(ref State state, bool compress)
```

Comment:
Probably want this internal, unless you think someone needs this for external use.